### ISSUE TITLE
Uploading the test pdf instead of scanning qr code

### ISSUE THREAD

**Issue by nicolone** [corona-warn-app/cwa-wishlist#389]
## Feature description

Uploading the test pdf instead of scanning

## Problem and motivation

I get the test result via email on my phone. I open the app and try to register my test result. But the app only allows to scan the test result qr code via the camera. So I need to open the pdf on a second device in order to be able to scan it. That is cumbersome.

## Is this something you're interested in working on

No (not an app developer)

**Comment by achisto** [comment]
I think the easiest solution is to enable CWA scanning QR-codes in pictures. Therefore a user could make a screenshot of their at-code (like in your PDF file) and import this picture to CWA.

**Comment by MikeMcC399** [comment]
@nicolone 

Thanks for raising this as a separate topic. I already suggested it in https://github.com/corona-warn-app/cwa-documentation/issues/400#issuecomment-785743476. Still waiting for feedback.

**Comment by dsarkar** [comment]
@nicolone Thanks for proposal. This is actually been discussed internally.

**Comment by TheFerryman** [comment]
What is the status of this issue? Any results on the internal discussions @dsarkar ?

**Comment by heinezen** [comment]
@TheFerryman Making QR code scanning more accessible is currently considered high priority. This includes this feature request as well as the one in https://github.com/corona-warn-app/cwa-wishlist/issues/302. These also are one of the most requested features right now.

However, we cannot give details about what the actual solution will look like yet. But we'll keep everyone here updated if we get morre news on this.

----
Corona-Warn-App Open Source Team

**Comment by muescha** [comment]
When i see the PDF in the iOS PDF View I would like to have the option to send it to the CWA App

**Comment by ndegendogo** [comment]
Of course the feasibility of such a feature depends on the details how the test centers provide the non-paper version of my QR code. This varies a lot across the test centers.
Also the feasibility and effort of such a feature need to be analysed separately for Android and iOS.

In the hope that a broader list of true use case scenarios helps the development team in their analysis, decision, and solution design, I give my own experience in more detail below:

- Date: 2021-04-22 I got a link via email
- link points to their website (https://...), and the URI references a file (.../image.png). My  code is still available today (which I didn't expect, tbh!) but maybe the server generates it on -the-fly from the given URI.
- I follow the link on my **iPhone**, and email opens the safari browser and shows the QR code.
- safari browser has a 'sharing' button. Here I can choose the format (automatic or image), and the destination for sharing (email, various messenger apps, copy (= for later paste), save (to a file or generic = the photo app), and a few more options). 
- Note that in context of sending it to email, "automatic" sends the link, whereas "image" sends the png (but this may differ between receiving applications, for Threema I got the image).

I don't know how my browser compiles this list of destination applications.

**But if cwa would succeed to register into this list and receive the CR code, this would be fantastic!**

**Comment by Ein-Tim** [comment]
FYI: @GisoSchroederSAP wrote in https://github.com/corona-warn-app/cwa-wishlist/issues/346#issuecomment-853875144:

> I'm totally aware of the difficulties that come with the **fulli-digitalized** process, where you don't even get something to scan, but just a link to a QR code or an email carrying a QR code.
> 
> We addressed this topic already multiple times. There is a slight chance, it becomes added _**if**_ (and only if) we are permitted to go for a face-lifting of the app. Main objective is to optimize the user experience, to streamline navigation and to make the UI more consistent on the way it works.

**Comment by dsarkar** [comment]
Edited 8.6.2021. 1700 due to comment https://github.com/corona-warn-app/cwa-wishlist/issues/389#issuecomment-856830615



**DISCLAIMER: The following is not an official recommendation, it is information on a possible workaround.  Privacy implications of using a cloud service have not been evaluated.**

Workaround works for event QR codes, for the moment I assume it will work for RAT QR codes, but not PCR codes (URL contains localhost, instead of s.coronawarn.app pr e.coronawarn.app) 

## Android

<img width="200" src="https://f-droid.org/repo/com.example.barcodescanner/en-US/phoneScreenshots/2_scan_from_file.png">

https://github.com/dmitriy-ilchenko/QrAndBarcodeScanner is an Android QR code scanner that is able to scan from images/screenshots. It has an Open Source License and is available in Play Store and on F-Droid.

@heinezen confirmed that scanning is possible without an internet connection.



<img width="1166" alt="image" src="https://user-images.githubusercontent.com/1409741/121204668-53f28c00-c877-11eb-893c-ab09114e4332.png">

This workaround uses the Google lens functionality.
Drawback: needs internet connection

## iOS

The google app for iOS has the Google lens functionality, however, it opens just the download page instead of the app. 
iOS 15 will include a similar native feature ("Live text")

--> Might be worth a FAQ.

**Comment by MikeMcC399** [comment]
@dsarkar 
Did you check on the privacy implications of using a cloud service for this?

**Comment by dsarkar** [comment]
@MikeMcC399 No, I didn't, thank you for asking! I will add a disclaimer in the post: This is not a recommendation, it is just information for the moment.

**Comment by heinezen** [comment]
@MikeMcC399 I added a more privacy friendly alternative for Android.

Don't know about iOS though. I don't think there are any scanners available that are Open Source, but correct me if I'm wrong.

----
Corona-Warn-App Open Source Team

**Comment by Ein-Tim** [comment]
One more reason why this should be implemented is that there are some users with non-working cameras (saw that on Twitter).
Since the app also doesn't allow to switch to the selfie camera there is no way for somebody with a broken back-camera to import a PCR test in the app at the moment.

**Comment by Ein-Tim** [comment]
Maybe we could adopt the title and the OP of this issue to be more general (remove "test") since this now also applies to vaccination certificates since some states are starting to send them out via E-Mail.

**Comment by jucktnich** [comment]
Maybe make the title more generic like "Scanning qr code from file"

**Comment by Ein-Tim** [comment]
@jucktnich 

I said the same in https://github.com/corona-warn-app/cwa-wishlist/issues/389#issuecomment-860382151

@dsarkar or @heinezen Could you maybe do this? Thanks!

**Comment by dsarkar** [comment]
Internal Tracking ID: [EXPOSUREAPP-5095](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-5095)
Internal Tracking ID: [EXPOSUREAPP-8358](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-8358) Android
Internal Tracking ID: [EXPOSUREAPP-8357](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-8357) iOS

**Comment by Ein-Tim** [comment]
Seems like they are working on this: https://github.com/corona-warn-app/cwa-app-android/pull/3837

**Comment by MikeMcC399** [comment]
It's good to see that the feature of scanning QR codes from a file is labelled with Fix 2.11!

**Comment by jucktnich** [comment]
Not for iOS 😢

**Comment by MikeMcC399** [comment]
@jucktnich 
> Not for iOS

I read that too:
https://github.com/corona-warn-app/cwa-app-ios/pull/3549#pullrequestreview-759464971

So it needs a new fix label for iOS, not 2.11.

**Comment by Ein-Tim** [comment]
Hey @nicolone!

Your suggestion has been implemented under Android with version 2.11:[Corona-Warn-App version 2.11 simplifies QR code scanning](https://www.coronawarn.app/en/blog/2021-10-06-cwa-version-2-11)

This version was released to the Stores today! 

Under iOS, users have to wait until the next version (2.12), then they can also use this feature. 

I suggest to leave this issue open until version 2.12 has been released.

**Comment by Ein-Tim** [comment]
Hello again @nicolone!

Now it is also possible under iOS to import QR codes from files.

Version 2.12 has been released & you can find more information in this blog post: [Corona-Warn-App version 2.12 enables central access to QR code scanner and expands statistics area](https://www.coronawarn.app/en/blog/2021-10-19-cwa-version-2-12)

I suggest closing this issue now.

**Comment by dsarkar** [comment]
Good morning @nicolone, @jucktnich, @MikeMcC399, @achisto, @TheFerryman, @Ein-Tim and community! Thanks for contributing here! Best wishes, DS 

---
Corona-Warn-App Open Source Team
